### DISCUSSION TITLE
HardwareSerialMIDI_Interface : Play note with ESP32

### DISCUSSION THREAD

**Discussion by U2Vi-2** [tttapa/Control-Surface#1002] [Q&A]
Hi everybody,

I want to run the lib on an ESP32-WROOM-32 with the physical MIDI pins.

Unfortunately, for the electronics engineer that I am, there are no electronic diagrams and the documentation is a horrible headache because it is so vast (but, I am sure that for a developer, this documentation is a gold mine :D ) .

So I'll try to take the time to understand all of this :)

First, I just want to play one note. I did the following wiring:

![Nouvelle-image-bitmap](https://github.com/tttapa/Control-Surface/assets/159192833/12b3751b-da0b-43dd-97a5-0e21de784262)

When pressing the button, I can get information about the serial port of the Arduino IDE:
```
Note On          Channel: 1	Data 1: 0x3c	Data 2: 0x7f
Note Off         Channel: 1	Data 1: 0x3c	Data 2: 0x7f
```
But the LED remains off and absolutely nothing happens on the RX2 and TX2 outputs.

```
/* Written by PieterP, 2020-01-09  
 * https://github.com/tttapa/Control-Surface */

#include <Control_Surface.h> // Include the Control Surface library

HardwareSerialMIDI_Interface midi {Serial2, MIDI_BAUD};
USBDebugMIDI_Interface midi2 = 115200;

// Instantiate a NoteButton object
NoteButton button {27,           // Push button on pin 27
  {MIDI_Notes::C(4), Channel_1}, // Note C4 on MIDI channel 1
};

// Instantiate the LED that will light up when middle C is playing
NoteLED led {26,                 // Pin of built-in LED
  {MIDI_Notes::C(4), Channel_1}, // Note C4 on MIDI channel 1
};

void setup() {
  Control_Surface.begin(); // Initialize the Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}
```
I wonder what I did wrong, I've been looking for the problem for a few hours in vain. I would be delighted if someone could help me tell me where I am going wrong :)

Thx,
Seb

- - -

Bonsoir,

Je souhaite faire fonctionner la lib sur un ESP32-WROOM-32 avec les broches MIDI physiques. 

Malheureusement, pour l'électronicien que je suis, il n'y a aucun schémas électronique et la documentation est une horrible prise de tête tellement elle est vaste (mais je suis sur que pour un développeur, cette documentation est une mine d'or).  

Je vais donc essayer de prendre le temps de comprendre tout ca :) 

Dans un premier temps, je souhaite simplement jouer une note. J'ai fais le câblage suivant : 

![Nouvelle-image-bitmap](https://github.com/tttapa/Control-Surface/assets/159192833/12b3751b-da0b-43dd-97a5-0e21de784262)

Lors de l'appuis sur le bouton, j'arrive bien à avoir des informations sur le port série de l'IDE Arduino : 
```
Note On          Channel: 1	Data 1: 0x3c	Data 2: 0x7f
Note Off         Channel: 1	Data 1: 0x3c	Data 2: 0x7f
```
Mais la led reste éteinte et il ne se passe strictement rien sur les sorties RX2 et TX2. 

```
/* Written by PieterP, 2020-01-09  
 * https://github.com/tttapa/Control-Surface */

#include <Control_Surface.h> // Include the Control Surface library

HardwareSerialMIDI_Interface midi {Serial2, MIDI_BAUD};
USBDebugMIDI_Interface midi2 = 115200;

// Instantiate a NoteButton object
NoteButton button {27,           // Push button on pin 27
  {MIDI_Notes::C(4), Channel_1}, // Note C4 on MIDI channel 1
};

// Instantiate the LED that will light up when middle C is playing
NoteLED led {26,                 // Pin of built-in LED
  {MIDI_Notes::C(4), Channel_1}, // Note C4 on MIDI channel 1
};

void setup() {
  Control_Surface.begin(); // Initialize the Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}
```

Je me demande ce que j'ai mal fais, cela fais quelques heures que je cherche le problème en vain. Je serais ravis qu'une personne puisse m'aider à me dire ou je fais fausse route :)

Bonne soirée,
Seb

**Comment by tttapa** [comment]
> ```cpp
> HardwareSerialMIDI_Interface midi {Serial2, MIDI_BAUD};
> USBDebugMIDI_Interface midi2 = 115200;
> ```

You're creating two MIDI interfaces here. Control Surface only uses the default MIDI interface. If you don't explicitly set the default, it's the one you've created last.

If you want to use multiple MIDI interfaces at the same time, see https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html#midi_md-routing and https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/da5/Dual-MIDI-Interface_8ino-example.html.

> But the LED remains off 

The NoteLED class only reacts to _incoming_ MIDI messages. The push button won't affect it directly.
